### ISSUE TITLE
yarn should run with --pure-lockfile

### ISSUE THREAD

**Issue by domoritz** [travis-ci/travis-ci#7395]
Without this option travis might overwrite the lockfile and it's hard to catch if the lockfile is out of date.

**Comment by BanzaiMan** [comment]
It seems best to override the default command in that case.

**Comment by BanzaiMan** [comment]
There was another person who wanted this as well.

@Daniel15 Do you have opinions on this?

**Comment by edmorley** [comment]
I was about to file an issue for this -- though I'd recommend using `--frozen-lockfile` instead, since it's closer to the desired effect (catch accidental config drift and quickly identify root cause).

The description from `yarn help install`...
```
    --no-lockfile                   don't read or generate a lockfile
    --pure-lockfile                 don't generate a lockfile
    --frozen-lockfile               don't generate a lockfile and fail if an update is needed
```

**Comment by BanzaiMan** [comment]
OK. We'll move to `--frozen-lockfile`.

**Comment by Daniel15** [comment]
This sounds reasonable to me. @bestander @kittens What do you think?

**Comment by bestander** [comment]
Agree

**Comment by BanzaiMan** [comment]
Reopening this since we've received some error reports and reverted the fix.

**Comment by BanzaiMan** [comment]
A use case https://github.com/travis-ci/travis-build/pull/997#issuecomment-288849044 raised was that Greenkeeper trying to update dependencies.

At the moment, the only flag we use for installing `yarn` (when it is not available on the image) is the presence of `yarn.lock`, so this puts such a use case in a conundrum. Though they can override `install` to overcome it, I think it is best to wait until all the build images have a recent version of `yarn` pre-installed.

**Comment by BanzaiMan** [comment]
On a second thought, the greenkeeper case is one where the user should override `install`. Eventually.

**Comment by BanzaiMan** [comment]
There was also a case where a pre-installed `yarn` flagged the build script to not install the latest version of `yarn`, where `--frozen-lockfile` was not available.

**Comment by madou** [comment]
Hey!

I'm getting an error `error: unknown option `--frozen-lockfile'` when running docker services:

```
services:
  - docker
```

Any ideas?

**Comment by ismay** [comment]
@madou So yarn is currently broken when also using docker services on travis? Because my builds are also failing with `error: unknown option --frozen-lockfile'`...

**Comment by bestander** [comment]
--frozen-lockfile was added around Yarn v0.20
What version do you use?

**Comment by ismay** [comment]
> What version do you use?

Just the version that travis installs automatically (when a yarn.lock is present)

**Comment by madou** [comment]
@ismay seems like it.

when I turned it off (see: https://travis-ci.org/madou/armory-back/jobs/214557773) it got past the install dependencies stage.

weird stuff!

**Comment by ismay** [comment]
@madou Yeah. Seems like `yarn --frozen-lockfile` is now run by default on travis builds for repos with a `yarn.lock`, but apparently with an outdated version of yarn which doesn't support that option?

Can't find a reference to the version of yarn that's being used in the build log though. Also weird that enabling docker makes it fail..

**Comment by madou** [comment]
in retrospect maybe its `sudo: true`. can't play with travis atm to find out though :-).

**Comment by BanzaiMan** [comment]
@ismay 
> Seems like yarn --frozen-lockfile is now run by default on travis builds for repos with a yarn.lock, but apparently with an outdated version of yarn which doesn't support that option?

See https://github.com/travis-ci/travis-build/pull/997 and the comments.

We have reverted this change, so restarting the failing jobs should work.

**Comment by madou** [comment]
@BanzaiMan do you need to do something special for the jobs that use `sudo`? because it still isn't working. 

see build I just ran: https://travis-ci.org/madou/armory-back/builds/214563501

**Comment by ismay** [comment]
@BanzaiMan @madou 

> We have reverted this change, so restarting the failing jobs should work.

Not for me, I've restarted the build several times, but it keeps failing. I'm not installing my own version of yarn, so the installation of yarn is being handled by travis: https://travis-ci.org/forecastxl/prototype/builds/214613995#L311

**Comment by BanzaiMan** [comment]
@madou @ismay I am so sorry about that. It is now deployed. Please restart your builds again.

**Comment by ismay** [comment]
@BanzaiMan [That did it for me](https://travis-ci.org/forecastxl/prototype/builds/214615030), thanks for the quick fix!

**Comment by madou** [comment]
Thanks @BanzaiMan !

**Comment by mmermerkaya** [comment]
Can you reconsider this?

More than a year has passed and the default Yarn version has supported `--frozen-lockfile` for a while now.

**Comment by bestander** [comment]
We are working out a solution for this, there should be another issue open

**Comment by mmermerkaya** [comment]
Thanks! There's a somewhat similar issue about NPM's lock file, but I commented here because I couldn't find any other open issue about this for Yarn.

**Comment by chocolateboy** [comment]
Related: https://travis-ci.community/t/run-yarn-with-frozen-lockfile-when-yarn-lock-is-present/3396
